### PR TITLE
8255883: Avoid duplicated GeneratedMethodAccessor when reflect method invoked from different threads

### PR DESCRIPTION
**Pull request by huishi-hs**
…AccessorImpl object

We met real problem when using protobuf with option optimized for code size, detail in JBS https://bugs.openjdk.java.net/browse/JDK-8255883

Optimize solution is adding a new boolean field to detect concurrent method accessor generation in same NativeMethodAccessorImpl object, only one thread is allowed to generate accessor, other threads still invoke in jni way until parent's delegator is updated from NativeMethodAccessorImpl  to generated accessor.

In common case, extra overhead is an atomic operation, compared with method accessor generate, this cost is trivial.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255883](https://bugs.openjdk.java.net/browse/JDK-8255883): Avoid duplicated GeneratedMethodAccessor when reflect method invoked from different threads

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to ec1531b26f6e1a311a2cef860d13522c09754305
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 85c8b96a1e4b20a881f04502bae8384223c68b62

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1070/head:pull/1070`
`$ git checkout pull/1070`

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java:80]: Why `succ` variable, if you can just `catch (Throwable e)` and restore the `accessorGenerated`?

File: src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java:66]: As the micro-optimization, checking that `accessor_generated` is `false` before attempting a (potentially contended) CAS might fit better. (See https://en.wikipedia.org/wiki/Test_and_test-and-set).

**Review by shipilev — COMMENTED** [review]
I do wonder if it makes sense to handle triple-state `int` here: "not yet generated", "generated", "in error"? So that we don't try to generate the accessor over and over again when it is in error?

File: src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java:80]: Good suggest! Will update.

File: src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java:66]: Agree this could benefit performance. Grep jdk codes (grep -R  compareAndSet *), not found codes with test and test-and-set pattern.

compareAndSet is intrinsic and maybe it's better let runtime/compiler generate Test_and_test-and-set code sequence to solve all compareAndSet operations.

**Comment by AlanBateman** [comment]
If we are changing NativeMethodAccessorImpl.invoke then we should probably do NativeConstructorAccessorImpl.newInstance at the same time. Also wondering if we should, while in the area, add "return acc.invoke(obj, args)" after setting the delegate so that it invokes the newly generated accessor.

Are there resource or other cases that you have observed where generateMethod fails and then succeeds in a subsequent call?

@cl4es Do you know of any startup tests that might be sensitive to the eager creating of a VarHandle?

I agree with @shipilev to test before the CAS.

**Comment by cl4es** [comment]
@AlanBateman I had a look and it seems unlikely the eager creation in this `<clinit>` will matter. I can find `NativeMethodAccessorImpl` in the startup profiles only for netty and larger apps, which often already initialize some `VarHandle` or are large enough that the initialization overhead will be lost in the noise.

**Comment by huishi-hs** [comment]
> If we are changing NativeMethodAccessorImpl.invoke then we should probably do NativeConstructorAccessorImpl.newInstance at the same time. 

Yes, NativeConstructorAccessorImpl should also apply this change.

> Also wondering if we should, while in the area, add "return acc.invoke(obj, args)" after setting the delegate so that it invokes the newly generated accessor.
> 

Agree,  I see no harm to invoke generated method when it is aviable. 
Should leave this to another patch?

> Are there resource or other cases that you have observed where generateMethod fails and then succeeds in a subsequent call?
> 
I have not seen exception/error happen in generate method yet.  But in case it fails in some ways, try - catch - reset is added to make sure behavior is same before/after with this change.

> @cl4es Do you know of any startup tests that might be sensitive to the eager creating of a VarHandle?
> 
> I agree with @shipilev to test before the CAS.

@AlanBateman 
Thanks for you comments!

**Comment by jbf** [comment]
Are there any benchmarks to compare this accessor with the previous version in the presumably common case where there is no or very little contention?

**Comment by huishi-hs** [comment]
@AlanBateman @shipilev  commit is updated, please help reivew it again

1.  add same change in NativeConstructorAccessorImpl
2.  eager creating of a VarHandle in both NativeConstructorAccessorImpl and NativeMethodAccessorImpl will cause error. [http://cr.openjdk.java.net/~hshi/8255883/cyclic_clinit_dependency.txt](url).  Use unsafe API here. 
3.  add test before compareAndSet and remove "succ" variable.

**Comment by huishi-hs** [comment]
> Are there any benchmarks to compare this accessor with the previous version in the presumably common case where there is no or very little contention? Edit to clarify: it is stated as "trivial" is this also measured somewhere?

@jbf  in presumably common case,  one extra CAS is added when invocation reach threshold.
MethodInvoke in jdk repo micro benchmark test shows no regression with this patch.

**Before patch**
Benchmark                                  Mode  Cnt   Score   Error  Units
MethodInvoke.invokeWithSixObjectParams     avgt   25   5.443 ? 0.008  ns/op
MethodInvoke.invokeWithSixPrimitiveParams  avgt   25  10.195 ? 0.020  ns/op
MethodInvoke.invokeWithoutParams           avgt   25   3.919 ? 0.011  ns/op

Benchmark                                  Mode  Cnt   Score   Error  Units
MethodInvoke.invokeWithSixObjectParams     avgt   25   5.440 ? 0.012  ns/op
MethodInvoke.invokeWithSixPrimitiveParams  avgt   25  10.197 ? 0.028  ns/op
MethodInvoke.invokeWithoutParams           avgt   25   3.923 ? 0.010  ns/op

**After patch**
Benchmark                                  Mode  Cnt   Score   Error  Units
MethodInvoke.invokeWithSixObjectParams     avgt   25   5.438 ? 0.005  ns/op
MethodInvoke.invokeWithSixPrimitiveParams  avgt   25  10.190 ? 0.038  ns/op
MethodInvoke.invokeWithoutParams           avgt   25   3.930 ? 0.010  ns/op
Finished running test 'micro:java.lang.reflect.MethodInvoke'

Benchmark                                  Mode  Cnt   Score   Error  Units
MethodInvoke.invokeWithSixObjectParams     avgt   25   5.442 ? 0.009  ns/op
MethodInvoke.invokeWithSixPrimitiveParams  avgt   25  10.205 ? 0.034  ns/op
MethodInvoke.invokeWithoutParams           avgt   25   3.916 ? 0.005  ns/op

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:37]: `static final` field identifiers should be capitalized. Suggestion: `private static final Unsafe U = Unsafe.getUnsafe()`.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:38]: Typo "Generated*a*Offset". But also this is `static final`, so it should be e.g. `GENERATED_OFFSET`. The field can be just `generated` to match the name of this offset.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:61]: Should be `!accessorGenerated`.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:62]: `compareAndSetBoolean` gets us into the awkward territory of sub-word CASes. (Click through to `Unsafe.compareAndExchangeByte` to see what I am talking about). It is probably not relevant here, though. Still, might be as good to use `int` field for generated flag.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (shipilev) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:72]: Typo: `Thowable`.

**Review by shipilev — CHANGES_REQUESTED** [review]
Minor nits here.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: fixed

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: field name udpated

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: change to "generated == 0", as generated is int type now

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: use int type now

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: updated

**Comment by huishi-hs** [comment]
@shipilev all comments are updated, please help review again!

**Review by shipilev — APPROVED** [review]
Looks okay (as far as low-level code goes) to me. Others need to ack too.

File: src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:44]: What is the reason for using an int? I remember there was a suggestion for three states but two states seems okay so curious why it was changed from boolean to int.
The restoring to 0 in the event of failure should probably be a volatile-write. Might be clearer to declare it as a volatile.

Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:44]: @AlanBateman 

> What is the reason for using an int? I remember there was a suggestion for three states but two states seems okay so curious why it was changed from boolean to int.

shipilev suggested not to use sub-word CAS, so change compareAndSetBoolean to compareAndSetInt. Change field "generated" from boolean to int doesn't increase object size (no extra byte/boolean to fold into same word).

> The restoring to 0 in the event of failure should probably be a volatile-write. Might be clearer to declare it as a volatile.
"generated" field  is read once and CAS once in this method. It doesn't likely cached and no visibiliy order required.

Comment (AlanBateman) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java:44]: okay, although I think it would be clearer for maintainers to declare it as volatile because it is set with CAS.

Comment (huishi-hs) [reply src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java]: @AlanBateman 

volatile attribute is added. As in common case, it only read and CAS at most once. And I see no regression on my testcase.

**Review by AlanBateman — APPROVED** [review]
Thanks for the update, latest version looks good.

**Comment by huishi-hs** [comment]
Thanks all! May this test-request get approved?

**Comment by JornVernee** [comment]
Notice that the `/test` command is currently un-available (some implementation concerns are still under consideration). As an alternative, GitHub actions can be used to do basic automatic testing when pushes occur.

However, pushes to the master branch are ignored. The tests didn't run for this PR since it was created from the `master` branch instead of a feature branch (this is ill-advised https://github.com/openjdk/jdk/pull/1070#issuecomment-722108502).

@huishi-hs My advice for now would be to push the changes for this PR to a separate new branch in your fork, for which the GitHub actions workflow should run, and then simply include a link to the workflow here (if you want to share the results).

**Comment by huishi-hs** [comment]
according to instructions, test is performed with another PR and passed https://github.com/openjdk/jdk/pull/1213

**Comment by huishi-hs** [comment]
@JornVernee How can "ready" label be added to this PR? This PR has been reviewed. Would you help on this?

**Comment by JornVernee** [comment]
@huishi-hs See the "Integration blocker" section in the PR body; you have a JBS issue and PR title mismatch.

**Comment by huishi-hs** [comment]
After review PR/commit title is changed to align with JBS title to add ready label

**Comment by AlanBateman** [comment]
Can you sync up your branch (the bot is saying that it's 140 commits behind the main line)? I can sponsor once it has been tested with an up to date branch.

**Comment by huishi-hs** [comment]
@AlanBateman Thanks for your sponsoring! Commit is updated.

**Pull request closed without merging**
